### PR TITLE
Tidy-up and latest CPython

### PR DESCRIPTION
**Pull request by Shillaker**
This started out as a catch-all for fixes to support ctypes properly, and it's just turned into a miscellaneous tidy-up:

- Refactored tasks around building cpython, copying libraries into place etc.
- Fixed the new ctypes-based pyfaasm to work in wasm as well (previously only worked natively with the emulator)
- Updated to latest version of our CPython fork which has some small changes to fix ctypes

### CODE REVIEW

File: libffi.imports [hunk 1]
```diff
@@ -1 +0,0 @@
-ffi_call
```
Comment (Shillaker) [inline libffi.imports:1]: All of libffi has been moved to faasm-toolchain

File: pyfaasm/pyfaasm/core.py [hunk 2]
```diff
@@ -7,62 +7,96 @@
...
-SUPPORTING_LIBS = [
+NATIVE_SUPPORTING_LIBS = [
     "libpistache.so",
     "libfaabricmpi.so",
 ]
 
-HOST_INTERFACE_LIB = "libemulator.so"
-PYTHON_LOCAL_CHAINING = bool(os.environ.get("PYTHON_LOCAL_CHAINING"))
-PYTHON_LOCAL_OUTPUT = bool(os.environ.get("PYTHON_LOCAL_OUTPUT"))
+NATIVE_INTERFACE_LIB = "libemulator.so"
```
Comment (Shillaker) [inline pyfaasm/pyfaasm/core.py:15]: This was previously written only for the native environment with the emulator, so the shared object loading has changed to be conditional

File: pyfaasm/pyfaasm/core.py [hunk 3]
```diff
@@ -7,62 +7,96 @@
...
 
 input_data = None
 output_data = None
 
 _host_interface = None
 
 
-def _init_host_interface():
-    global _host_interface
+def get_env_bool(var_name):
```
Comment (Shillaker) [inline pyfaasm/pyfaasm/core.py:25]: These are just convenience functions for handling flags from the environment

File: pyfaasm/pyfaasm/core.py [hunk 4]
```diff
@@ -7,62 +7,96 @@
...
+            _host_interface = ctypes.CDLL(NATIVE_INTERFACE_LIB)
+
+
+def get_input_len():
+    _init_host_interface()
+    return _host_interface.__faasm_read_input(None, 0)
 
 
-def read_input():
+def read_input(input_len):
```
Comment (Shillaker) [inline pyfaasm/pyfaasm/core.py:82]: The rest of the changes in this file are bug fixes discovered when testing the wasm interface

File: tasks/cpython.py [hunk 5]
```diff
@@ -39,6 +40,17 @@
...
+WASM_INCLUDES_DIR = join(WASM_SYSROOT, "include")
+
+LIB_SRC_DIR = join(INSTALL_DIR, "lib")
+LIB_DEST_DIR = join(FAASM_RUNTIME_ROOT, "lib")
+
+LIBPYTHON_SRC_PATH = join(LIB_SRC_DIR, "libpython3.8.a")
+LIBPYTHON_DEST_PATH = join(WASM_LIB_INSTALL, "libpython3.8.a")
+
+INCLUDE_SRC_DIR = join(INSTALL_DIR, "include", "python3.8")
+INCLUDE_DEST_DIR = join(WASM_INCLUDES_DIR, "python3.8")
```
Comment (Shillaker) [inline tasks/cpython.py:52]: The changes in this file are to copy the cpython build outputs into place in `/usr/local/faasm`. Previously this was done in the Docker build, and meant you couldn't easily develop using this and faasm in the same container

File: tasks/runtime.py [hunk 6]
```diff
@@ -1,21 +1,26 @@
...
+from shutil import rmtree
 from subprocess import run
 from os import makedirs, remove
 
 from invoke import task
 
-from faasmtools.build import FAASM_LOCAL_DIR
+from tasks.env import PROJ_ROOT, THIRD_PARTY_DIR, FAASM_RUNTIME_ROOT
 
-from tasks.env import PROJ_ROOT, THIRD_PARTY_DIR
```
Comment (Shillaker) [inline tasks/runtime.py:12]: This task used to copy cpython into place as well as the runtime files, but the copying of cpython has been moved to the `cpython` task

File: tasks/runtime.py [hunk 7]
```diff
@@ -1,21 +1,26 @@
...
 CROSSENV_WASM_DIR = join(PROJ_ROOT, "cross_venv", "cross")
 
+LIB_SRC_DIR = join(CPYTHON_INSTALL_DIR, "lib", "python3.8")
+LIB_DEST_DIR = join(FAASM_RUNTIME_ROOT, "lib", "python3.8")
+
+SITE_PACKAGES_SRC_DIR = join(
+    CROSSENV_WASM_DIR, "lib", "python3.8", "site-packages"
+)
+SITE_PACKAGES_DEST_DIR = join(LIB_DEST_DIR, "site-packages")
+
```
Comment (Shillaker) [inline tasks/runtime.py:23]: Factoring everything out into these variables led to a large diff for what is really a small change, sorry.

File: tasks/cpython.py [hunk 8]
```diff
@@ -66,7 +78,7 @@ def build(ctx, clean=False, noconf=False, nobuild=False):
     # relevant in the module builds.
 
     # Link in extra wasi-libc long double support (see wasi-libc docs)
-    link_libs = "-lc-printscan-long-double"
+    link_libs = "-lc-printscan-long-double -lfaasm"
```
Comment (Shillaker) [inline tasks/cpython.py:81]: This is to support using Faasm functions like backtrace in cpython if need be

**Review by csegarragonz — APPROVED** [review]

**Pull request merged**